testing/e2e/describes-e2e.js

```diff
@@ -0,0 +1,200 @@
+   * @param {!Object} spec
+   * @param {function(!Object)} fn
+   * @param {function(string, function())} describeFunc
+   */
+  const templateFunc = function(name, spec, fn, describeFunc) {
+    const fixtures = [];
+    factory(spec).forEach(fixture => {
+      if (fixture && fixture.isOn()) {
+        fixtures.push(fixture);
+      }
+    });
+    return describeFunc(name, function() {
+      const env = Object.create(null);
+      this.timeout(TIMEOUT);
+      beforeEach(() => {
+        let totalPromise = undefined;
+        // Set up all fixtures.
+        fixtures.forEach((fixture, unusedIndex) => {
+          if (totalPromise) {
+            totalPromise = totalPromise.then(() => fixture.setup(env));
+          } else {
+            const res = fixture.setup(env);
+            if (res && typeof res.then == 'function') {
+              totalPromise = res;
+            }
+          }
+        });
+        return totalPromise;
+      });
+
+      afterEach(function() {
+        // Tear down all fixtures.
+        fixtures.slice(0).reverse().forEach(fixture => {
+          // TODO(cvializ): handle errors better
+          // if (this.currentTest.state == 'failed') {
+          //   fixture.handleError();
+          // }
+          fixture.teardown(env);
+        });
+
+        // Delete all other keys.
+        for (const key in env) {
+          delete env[key];
+        }
+      });
+
+      describe(SUB, function() {
+        fn.call(this, env);
+      });
+    });
+  };
+
+  /**
+   * @param {string} name
+   * @param {!Object} spec
+   * @param {function(!Object)} fn
+   */
+  const mainFunc = function(name, spec, fn) {
+    return templateFunc(name, spec, fn, describe);
+  };
+
+  /**
+   * @param {string} name
+   * @param {!Object} spec
+   * @param {function(!Object)} fn
+   */
+  mainFunc.only = function(name, spec, fn) {
+    return templateFunc(name, spec, fn, describe./*OK*/only);
+  };
+
+  mainFunc.skip = function(name, variants, fn) {
+    return templateFunc(name, variants, fn, describe.skip);
+  };
+
+  return mainFunc;
+}
+
+
+/** @interface */
+class FixtureInterface {
+
+  /** @return {boolean} */
+  isOn() {}
+
+  /**
+   * @param {!Object} env
+   * @return {!Promise|undefined}
+   */
+  setup(unusedEnv) {}
+
+  /**
+   * @param {!Object} env
+   */
+  teardown(unusedEnv) {}
+}
+
+/** @implements {FixtureInterface} */
+class AmpPageFixture {
+
+  /** @param {!TestSpec} spec */
+  constructor(spec) {
+    /** @const */
+    this.spec = spec;
+
+    // /** @private @const {!Array<string>} */
+    // this.browsers_ = this.spec.browsers || ['chrome'];
+
+    /** @private @const */
+    this.driver_ = null;
+  }
+
+  /** @override */
+  isOn() {
+    return true;
+  }
+
+  /** @override */
+  setup(env) {
+    // TODO(esth): implement sessions
+    // TODO(esth): ensure tests are in a sandbox
+    // See https://w3c.github.io/webdriver/#sessions
+
+    // TODO(esth): create multiple drivers per 'config.browsers'
```

Comment:
If you've already filed separate issues for these items, mention the issue numbers in the `TODO`s:
```
TODO(estherkim, 123456): Do so-and-so.
```